# Practice Midterm 2
### Psych 205 Practice Midterm 2

# NOTE: I WOULD USE THE QUICK REFERENCE INSTEAD. THE LAST QUESTION ON THIS PRACTICE MIDTERM IS INCORRECT

In [1]:
require(car)
require(stats)

Loading required package: car
Warning message:
"package 'car' was built under R version 3.4.3"

In [2]:
tooth = ToothGrowth

## Exercise 1. (1 point).  
### Visualize the data. Use the scatterplot command to make a scatterplot of the data with one line for each of the delivery methods. Perform a two-way anova (with interaction) to evaluate 


In [3]:
#scatterplot(len~dose|supp, data=ToothGrowth)


In [4]:
table(tooth$supp)


OJ VC 
30 30 

In [5]:
table(tooth$dose)


0.5   1   2 
 20  20  20 

In [6]:
summary(tooth$len)

   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   4.20   13.07   19.25   18.81   25.27   33.90 

In [7]:
#Two way anova with interaction effect (http://www.sthda.com/english/wiki/two-way-anova-test-in-r)
summary(aov(len ~ supp*dose, data=ToothGrowth))

            Df Sum Sq Mean Sq F value   Pr(>F)    
supp         1  205.4   205.4  12.317 0.000894 ***
dose         1 2224.3  2224.3 133.415  < 2e-16 ***
supp:dose    1   88.9    88.9   5.333 0.024631 *  
Residuals   56  933.6    16.7                     
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

# Exercise 2
Test the predictive power of three nested models 
* that correspond to a single line, 
* two lines with different intercepts and 
* two lines with different intercepts and slopes.

Type in the commands to make these models and using the summary() command comment on the results.  
You comments should address both the values of the model coefficients and of the measures of goodness of fit.
Type in your comments as "comments" in this R script below the R commands.

### Single line

In [8]:
mod1 = lm(len ~ dose, data=tooth)
summary(mod1)
#R^2 and adjR2 of .64 and .63 respectively


Call:
lm(formula = len ~ dose, data = tooth)

Residuals:
    Min      1Q  Median      3Q     Max 
-8.4496 -2.7406 -0.7452  2.8344 10.1139 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)   7.4225     1.2601    5.89 2.06e-07 ***
dose          9.7636     0.9525   10.25 1.23e-14 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 4.601 on 58 degrees of freedom
Multiple R-squared:  0.6443,	Adjusted R-squared:  0.6382 
F-statistic: 105.1 on 1 and 58 DF,  p-value: 1.233e-14


### Two lines, different intercept

In [9]:
mod2 = lm(len ~ supp + dose, data=tooth)
summary(mod2)
#both R2 and adjR2 increase massivly here -- going to .7 and .69 respectively
# This indicates that adding dose means our model  explains much more of the
#variability in length of teeth
#and both are higher than the single term model, so we aren't being punished so much
#for additional parameters that this isn't a benefit - it is
#Both dose and supp are significant, dose extremely significant with a much higher 
#coefficient. This seems to imply that dose was a very important term for doing 
#accurate predictions
#the reference level appears to be Supp OJ 
#VC decreases tooth length relative to OJ all else equal
# increased dose increases tooth length


Call:
lm(formula = len ~ supp + dose, data = tooth)

Residuals:
   Min     1Q Median     3Q    Max 
-6.600 -3.700  0.373  2.116  8.800 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)   9.2725     1.2824   7.231 1.31e-09 ***
suppVC       -3.7000     1.0936  -3.383   0.0013 ** 
dose          9.7636     0.8768  11.135 6.31e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 4.236 on 57 degrees of freedom
Multiple R-squared:  0.7038,	Adjusted R-squared:  0.6934 
F-statistic: 67.72 on 2 and 57 DF,  p-value: 8.716e-16


### Two lines, different slopes

In [10]:
mod3 = lm(len ~ supp*dose, data=tooth)
summary(mod3)
#R2 and adjR2 increased - despite additional terms, it is still higher than the
#model, meaning more variability is explained even when adjusted for the additional 
#terms. All terms are significant, suggesting they are related
#the reference level appears to be Supp OJ 
#taken as a whole, VC decreases tooth length relative to OJ all else equal
# interacting with dose, VC has an estimate of 3.9 as opposed to the normal dose
#slope of 7.8 - implying again that even in the interaction VC is providing less of 
#an increase in length as dose increases
#dose is positively related to tooth length. (same description as in previous q)

#Comparing adjusted R^2, it increases from model1 -> model2 -> model3. 
#So, it appears that even with the additional parameters, adding these terms is still
#making this regression more accurate



Call:
lm(formula = len ~ supp * dose, data = tooth)

Residuals:
    Min      1Q  Median      3Q     Max 
-8.2264 -2.8462  0.0504  2.2893  7.9386 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)   11.550      1.581   7.304 1.09e-09 ***
suppVC        -8.255      2.236  -3.691 0.000507 ***
dose           7.811      1.195   6.534 2.03e-08 ***
suppVC:dose    3.904      1.691   2.309 0.024631 *  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 4.083 on 56 degrees of freedom
Multiple R-squared:  0.7296,	Adjusted R-squared:  0.7151 
F-statistic: 50.36 on 3 and 56 DF,  p-value: 6.521e-16


# Exercise 3.
How many parameters are there in each of the three models?
Make a line plot of model parameters versus R2adj. You can get R2adj from the summary commands.

In [11]:
m1len = length(coef(mod1))
#2
m2len = length(coef(mod2))
#3
m3len = length(coef(mod3))
#4
# get adjR2
m1r2 = summary(mod1)$adj.r.squared
m2r2 = summary(mod2)$adj.r.squared
m3r2 = summary(mod3)$adj.r.squared
#plot
#plot(c(m1len,m2len,m3len),c(m1r2,m2r2,m3r2), type="l")

# Exercise 4
Calculate the R2, R2 adj, F and p value for the full model "by hand".
Make sure that the results match those you obtained from the summary command.

## GOALS
Multiple R-squared:  0.7296,	Adjusted R-squared:  0.7151 
F-statistic: 50.36 on 3 and 56 DF,  p-value: 6.521e-16

## R2

In [12]:
mod0 = lm(len ~ 1, data=tooth)
tooth_hat_full = predict(mod3)
tooth_hat_null =  predict(mod0)

ss3 = sum((tooth$len - tooth_hat_full)^2)
ss0 = sum((tooth$len - tooth_hat_null)^2)

R2 = 1 - (ss3/ss0)
print(R2)

[1] 0.7295544


## Adjusted R2

In [13]:
n = length(mod3$model$len)
k = length(mod3$coefficients)

R2adj = 1 - (ss3/(n-k))/(ss0/(n-1))
print(R2adj)

[1] 0.7150662


## F value

In [14]:
F_val = ((ss0 - ss3)/(k-1))/((ss3)/(n-k))
print(F_val)

[1] 50.35522


## P value

In [15]:
require(stats)
pfval = 1 - pf(F_val, df1 = (k-1), df2 = n-k)

# Exercise 5
Using 10 fold cross-validation estimate the cross-validated
R2 for the three models.  Make a plot of average R2CV versus number of parameters.

In [16]:
df = tooth

#find the number of folds
n_folds = 10

#get rid of nas, shuffle data
df <-na.omit(df[sample(nrow(df)),])

#create folds
folds <- cut(seq(1,nrow(df)),breaks=n_folds,labels=FALSE)

#create arrays for results
R2cv_mod1 = array(data=0, dim=n_folds)
R2cv_mod2 = array(data=0, dim=n_folds)
R2cv_mod3 = array(data=0, dim=n_folds)

#for loop
for(i in 1:n_folds){
  
  #split up data
  test = df[folds==i,]
  train = df[folds !=i,]
  
  #fit models
  mod0 = lm(len ~ 1, data=train)
  mod1 = lm(len ~ supp, data=train)
  mod2 = lm(len ~ supp + dose, data=train)
  mod3 = lm(len ~ supp*dose, data=train)
  
  yhat_mod0 = predict(mod0, data=test)
  yhat_mod1 = predict(mod1, data=test)
  yhat_mod2 = predict(mod2, data=test)
  yhat_mod3 = predict(mod3, data=test)
  
  ss0 = sum((yhat_mod0 - test$len)^2)
  ss1 = sum((yhat_mod1 - test$len)^2)
  ss2 = sum((yhat_mod2 - test$len)^2)
  ss3 = sum((yhat_mod3 - test$len)^2)
  
  R2cv_mod1[i] = ss1/ss0
  R2cv_mod2[i] = ss2/ss0
  R2cv_mod3[i] = ss3/ss0
  
}

meanR2_1 = mean(R2cv_mod1)
meanR2_2 = mean(R2cv_mod2)
meanR2_3 = mean(R2cv_mod3)

means = c(meanR2_1, meanR2_2, meanR2_3)
params = c(m1len, m2len, m3len)

#plot(params, means, type='l')

# Exercise 6.
First, use the appropriate classical commands to statiscally compare the single line model to
 the full model. 
 
 Second, using the results of the cross-validation in 5, determined whether you would reach the same conclusion. 
 
 For example compare the mean R2cv plus or minus two standard errors to assess whether one is significantly
 better.

### Traditional methods

In [17]:
anova(mod1, mod3)

Res.Df RSS       Df Sum of Sq F        Pr(>F)      
1 52     2663.3760 NA       NA        NA           NA
2 50      803.7729  2 1859.603  57.83981 9.829296e-14

In [18]:
SSfull = sum((mod3$residuals)^2)
SSnull = sum((mod1$residuals)^2)

kfull = length(mod3$coefficients)
ksimple = length(mod1$coefficients)

n = length(mod3$model$len)

#calculating difference in quality using R2 and R2adjusted
R2 = 1-SSfull/SSnull #atypical way to do this
print(R2)

R2adj = 1-(SSfull/(n-kfull))/(SSnull/(n-ksimple))
print(R2adj)


#diff method 2
R2change = summary(mod3)$r.squared - summary(mod1)$r.squared
print(R2change)

#F val and p val

q = kfull - ksimple

Fval = ((SSnull - SSfull)/q)/(SSfull/(n-kfull))

pval = pf(Fval, df1=q, df2 = n-k, lower.tail=FALSE)


[1] 0.6982127
[1] 0.6861413
[1] 0.6316379


In [19]:
Fval

[1] 57.83981

In [20]:
pval

[1] 9.829296e-14

According to these results, model 3 does work significantly better than model 1. The F value of 58 with a p-value of <.00001 implies that model 3 is significantly better than model 1, as that F value would be incredibly unlikely under the null hypothesis tha the model peformance is identical

### Non-traditional methods

In [21]:
summary(R2cv_mod1)

   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 0.9639  1.0104  1.0820  1.0905  1.1110  1.3495 

In [22]:
summary(R2cv_mod3)

   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  1.400   1.528   1.757   1.805   2.136   2.240 

In [26]:
se_mod1 = sd(R2cv_mod1)/sqrt((10))

In [29]:
se_mod3 = sd(R2cv_mod3)/sqrt((10))

In [31]:
se_mod1

[1] 0.03502614

In [32]:
se_mod3

[1] 0.105388

In [34]:
(meanR2_3 - se_mod3*2)>(meanR2_1 - se_mod1*2)

[1] TRUE